[PR-1427] Structs in Rust

sancho20021 opened the pull request.

sancho20021 committed:
[PR-1427] Write blog post

NaeosPsy commented on structs-in-rust/article.md:
````diff
@@ -0,0 +1,366 @@
+```
+struct A;
+```
+Don't be afraid if you see such strange struct definitions. You may wonder what is the purpose of unit structs when they don't contain any information. In reality, they are being used (see [std::fmt::Error](https://doc.rust-lang.org/std/fmt/struct.Error.html), [chrono::offset::Utc](https://docs.rs/chrono/0.4.19/chrono/offset/struct.Utc.html)), as they can have methods and different trait implemen ...
````
This isn't very convincing, could you come up with a simple example?

sancho20021 replied:
Ok. I think the example would require generics and of course traits. What do you think of placing the example after the `Traits` section?

NaeosPsy replied:
That feels too complicated for this. 🤔 Can it be textual -- a description of what would you do.

sancho20021 replied:
I want to show smth like this:
```
mod library {
    pub trait ErrorLogger {
        fn error(error_code: i32);
    }

    pub fn function_with_errors<L: ErrorLogger>() {
        L::error(25);
    }
}

mod my_app {
    use super::library::*;

    // Logger 1
    struct PoliteLogger;

    impl ErrorLogger for PoliteLogger {
        fn error(error_code: i32) {
            println!(
                "Sorry for interrupting you, but an error with code {} occurred.",
                error_code
            );
        }
    }

    // Logger 2
    struct SimpleLogger;

    impl ErrorLogger for SimpleLogger {
        fn error(error_code: i32) {
            println!("Error: {}", error_code);
        }
    }

    fn option1() {
        function_with_errors::<PoliteLogger>();
    }

    fn option2() {
        function_with_errors::<SimpleLogger>();
    }
}
```

So do I need to describe the idea in words? But we assume the reader doesn't know that functions can take generic arguments.

sancho20021 replied:
What about dropping unit structs? I can't come up with a simple explanation without leaving the reader with questions

NaeosPsy replied:
Yeah, that is possible as well. 👍

sancho20021 replied:
I decided to remove it)

NaeosPsy commented on structs-in-rust/article.md:
```diff
@@ -0,0 +1,366 @@
+- Have an immutable binding to the struct. Then you cannot change any of its fields.
+- Have a mutable binding, in which case you can set whatever field you choose.
+
+### Convenient field init syntax
```
Wait, how does this differ from regular initialization?

sancho20021 replied:
Ah, I see, misleading header. I made a change in the latest commit which should fix the problem

NaeosPsy commented on structs-in-rust/article.md:
````diff
@@ -0,0 +1,366 @@
+It is possible to omit *field* in `field: value` initialization syntax if the value is a variable (or function argument) with a name that matches the field:
+```
+fn new_bicycle(brand: String) -> Bicycle {
+    let kind = String::from("road");
````
I'd move this to the argument, it's simpler that way.

sancho20021 replied:
Ok

NaeosPsy committed:
Quick preliminary edits

NaeosPsy commented on structs-in-rust/article.md:
```diff
@@ -0,0 +1,419 @@
+
+The most important lines in this snippet are those where we create `point` and `double_bool`. Because both `Point` and `DoubleBool` implement `Default`, we can invoke the `default()` method of the `Default` trait in each case to generate an empty value of type `Point` and `DoubleBool` respectively.
+
+*Exercise: try to remove explicit type annotations from lines 3 and 4 and see what the compiler tells you.*
```
I'd add the word "Why?" at the end to make it feel more like an exercise.

sancho20021 replied:
Ok

NaeosPsy commented on structs-in-rust/article.md:
```diff
@@ -0,0 +1,419 @@
+The most important lines in this snippet are those where we create `point` and `double_bool`. Because both `Point` and `DoubleBool` implement `Default`, we can invoke the `default()` method of the `Default` trait in each case to generate an empty value of type `Point` and `DoubleBool` respectively.
+
+*Exercise: try to remove explicit type annotations from lines 3 and 4 and see what the compiler tells you.*
+
```
It seems that you could show how to manually implement a trait here as well, seems like there is space?

sancho20021 replied:
@NaeosPsy But not defining our own traits? Just implementing exising?

NaeosPsy replied:
Yeah.

sancho20021 replied:
I added a small example

sancho20021 committed:
Fix `Convenient field init syntax` section

sancho20021 committed:
Enhance `Excercise` description.

sancho20021 committed:
Add manual trait implementation

sancho20021 committed:
Remote unit struct

Vlad-Shcherbina commented on structs-in-rust/article.md:
```diff
@@ -0,0 +1,443 @@
+You may have noticed that methods have `self` as their first argument. It represents the instance of the struct the method is being called on, similar to how it is done in Python. Some syntax sugar is involved here. Let's desugar it in two steps.
+
+|     |                                              |                                             |
+| --- | -------------------------------------------- | ------------------------------------------- |
```
(optional) For completeness, maybe also desugar
```
// a method that consumes struct instance
fn foo(self)
```

Or did you omit it intentionally?

sancho20021 replied:
I think I omitted it because:
- I didn't show examples with `self`
- Don't want to explain `self` vs `&self`

Vlad-Shcherbina commented on structs-in-rust/article.md:
````diff
@@ -0,0 +1,443 @@
+let y = p_named.y;
+```
+
+One may wonder: what is the syntax to address a specific field of a tuple struct? Fortunately, Rust has chosen the simplest way possible, by enumerating tuple components from *0* to *n*.
````
Don't know what you mean by *n*, but maybe should be *0* to *n - 1*.

sancho20021 replied:
Rephrased

Vlad-Shcherbina commented on structs-in-rust/article.md:
````diff
@@ -0,0 +1,443 @@
+}
+```
+
+There are two options:
````
(this comment is optional; totally up to you if it's worth mentioning)

There is a third option, kind of: make the field private and don't expose methods that modify it.
It will be still possible to change that field by assigning to the whole struct.

sancho20021 replied:
Okay, I'll leave it as it is, but thanks for the idea, didn't think about it!

Vlad-Shcherbina commented on structs-in-rust/article.md:
```diff
@@ -0,0 +1,443 @@
+
+- [`Debug`](https://doc.rust-lang.org/std/fmt/trait.Debug.html). Used to format (and print) a value using `{:?}`.
+- [`Clone`](https://doc.rust-lang.org/std/clone/trait.Clone.html). Used to get a duplicate of a value.
+- [`Default`](https://doc.rust-lang.org/std/default/trait.Default.html). Used to create a default instance of type.
```
I'm not sure about `Default`. All other traits here are very important. `Default` is really not.

Is it worth bringing it up?

sancho20021 replied:
Removed

Vlad-Shcherbina commented on structs-in-rust/article.md:
````diff
@@ -0,0 +1,443 @@
+assert_eq!(format!("{}", p), String::from("(0, 0)"));
+```
+
+The syntax seems pretty natural. We `impl`ement the trait `std::fmt::Display` `for` the struct called `Point`. In the opened code block we then have to define every function that is declared in the trait. In our case, there are only one — `fmt`.
````
I think there is a general style advice to avoid phrases like "it's easy to see, that...".
If it's easy to see, the reader will notice that it's easy to see themselves.
If it's not, it's a lie.

Maybe drop "the syntax seems pretty natural" and proceed directly to the explanation.

NaeosPsy replied:
I agree. 👍

sancho20021 replied:
Yes, thanks for the advice!

Vlad-Shcherbina commented on structs-in-rust/article.md:
````diff
@@ -0,0 +1,443 @@
+```
+#### Set
+
+It is possible to change the value of a field using the same dot notation, but the struct instance **must be defined as mutable**.
````
Maybe "struct variable" instead of "struct instance"?
"Struct instance" sounds like a value in memory, and mutability is a property of bindings and references, not values.

sancho20021 replied:
Done

Vlad-Shcherbina commented on structs-in-rust/article.md:
```diff
@@ -0,0 +1,443 @@
+let p = Point::new(0, 1);
+
+// format a pair
+assert_eq!(format!("{:?}", p), String::from("Point { x: 0, y: 0 }"));
```
You can compare `String` with `&str`, so it could be simply
```
assert_eq!(format!("{:?}", p), "Point { x: 0, y: 0 }");
```

sancho20021 replied:
Ok

Vlad-Shcherbina commented on structs-in-rust/article.md:
```diff
@@ -0,0 +1,443 @@
+        Point { x, y }
+    }
+
+    // Immutable method
```
I don't think terms "immutable method" and "mutable method" are common. Sounds kind of inaccurate, I dunno.
I wish there were succinct phrases for that, but nothing comes to mind.

If we are trying to avoid the term "borrow" for simplicity, the clunky alternative is "a method that can't modify struct instance" and "a method that can modify struct instance".

sancho20021 replied:
Rephrased as suggested

Vlad-Shcherbina commented on structs-in-rust/article.md:
```diff
@@ -0,0 +1,443 @@
+There are at least two reasons.
+
+1. Nice-looking dot notation.
+2. Code organization. It is a lot more convenient for the user of your library to see all functionality of the struct inside one impl block instead of tens of functions spread all across the module.
```
I agree it's about code organization, but disagree with the explanation. Nothing prevents one from keeping regular functions together textually, or from scattering methods across multiple impl blocks all over:
```
struct MyStruct {}

...

impl MyStruct {
  fn do_stuff
}

...

impl MyStruct {
  fn do_other_stuff()
}
```

--------

There two things:
* You don't have to import methods:
    ```
    use my_module::MyStruct;

    ...
    let s = MyStruct::new();
    s.do_stuff();
    s.do_other_stuff();
    ```
    vs
    ```
    use my_module::{MyStruct, do_stuff, do_other_stuff};  // ugly

    ...
    let s = MyStruct::new();
    do_stuff(&s);
    do_other_stuff(&s);
    ```
* Methods (and associated functions in general) are namespaced. So you don't have name collisions and can write
    ```
    my_rectangle.area()
    my_circle.area()
    ```
    instead of more verbose and redundant
    ```
    rectangle_area(my_rectangle)
    circle_area(my_circle)
    ```

sancho20021 replied:
Done

Vlad-Shcherbina commented on structs-in-rust/article.md:
```diff
@@ -0,0 +1,443 @@
+
+### Why use methods instead of regular functions?
+
+There are at least two reasons.
```
Another tiny advantage of methods is autoreferencing:
You can write
```
point.shift_right(3);
```
instead of explicit
```
(&mut point).shift_right(3);
```

For functions, you always have to reference:
```
shift_point_right(&mut point, 3);
```

sancho20021 replied:
Added

Vlad-Shcherbina commented on structs-in-rust/article.md:
```diff
@@ -0,0 +1,443 @@
+1. Nice-looking dot notation.
+2. Code organization. It is a lot more convenient for the user of your library to see all functionality of the struct inside one impl block instead of tens of functions spread all across the module.
+
+## Methods vs traits
```
I dunno, is it worth going into this in a beginner-level post?

A beginner is unlikely to create their own traits, so the rule is simple:
* if you are implementing a common functionality for which a trait exists (converting to string, comparison, ...), maybe `impl` that trait
* if you are doing something specific to your application, probably use methods in regular `impl` blocks

NaeosPsy replied:
I think the rule is good and can be added.

The rest of it should probably stay, as much for beginners as for SEO.

sancho20021 replied:
Added

Vlad-Shcherbina reviewed (commented):
Overall this seems correct, writing style is precise and simple.

All my comments are optional, you are the author.

sancho20021 committed:
Fixes after Vlad's review

NaeosPsy committed:
Copy edit changes

I've added copy edits. Please check if they don't break anything. They do break the "methods" section -- I invite you to fix it by fixing the TODO's that I've left.

NaeosPsy commented:
@sancho20021 I've added copy edits. Please check if they don't break anything. They do break the "methods" section -- I invite you to fix it by fixing the TODO's that I've left.

sancho20021 committed:
[PR-1427] split methods and associated functions

sancho20021 committed:
[PR-1427] Move chapter about `Self` up

sancho20021 commented:
@NaeosPsy I've fixed the TODO's. Fixes are in the first commit.

Please, make sure you review the part about `Self`, because I rewritten the sentence.

And I also suggest a small reordering in the last commit, moving the chapter about `Self` up, right after the `&mut self` part, so the reader gets the explanation immediately. What do you think?

NaeosPsy committed:
Small fix

NaeosPsy commented:
Ok @sancho20021 , looks great, thank you very much! 🙏

I'll schedule it to be published next Tuesday, 16th August.

NaeosPsy committed:
Smallest of fixes

sancho20021 commented:
Great, thank you too!
